@font-face src format() should parse valid unsupported keywords

vitorroriz opened the pull request:
#### b2a00a22e3c413618df230613ed43dfbf9d2e39f
<pre>
@font-face src format() should parse valid unsupported keywords
<a href="https://bugs.webkit.org/show_bug.cgi?id=259144">https://bugs.webkit.org/show_bug.cgi?id=259144</a>
rdar://112135869

Reviewed by Tim Nguyen.

A valid format() keyword (ident) argument should be parsed even if the represented
format is not supported by the engine. The engine should reject loading
the font for unsupported formats but this should happen only at
loading time and not at parsing time.

The reason why we were rejecting it before, at parsing time, is for some
inconsistency in different parts of the spec. That
will be fixed soon by the CSSWG, see: <a href="https://github.com/w3c/csswg-drafts/issues/8793">https://github.com/w3c/csswg-drafts/issues/8793</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-format-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::isFontFormatKeywordValid):
(WebCore::CSSPropertyParserHelpers::consumeFontFormat):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::consumeSupportsFontFormatFunction):

Canonical link: <a href="https://commits.webkit.org/266043@main">https://commits.webkit.org/266043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a9aa35b979ae93ecdfdedf7a6b5b1af37dcc815

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13297 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14392 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12112 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12714 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12999 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14812 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13591 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10711 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14843 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10868 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11456 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11622 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14828 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10012 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11345 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3106 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11954 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -8256,12 +8272,11 @@ String consumeFontFormat(CSSParserTokenRange& range, bool rejectStringValues)
-    return format;
+    if (arg.type() == IdentToken && isFontFormatKeywordValid(arg.id()))
+        return arg.value().toString();
+    if (arg.type() == StringToken && !rejectStringValues)
```
Do you know if strings should reject keywords that don't match `isFontFormatKeywordValid` ?

vitorroriz replied:
I don't think so, since the syntax takes `<String>`, any string should be parsed as a valid format, and then we should reject it at loading time.

nt1m replied:
Ah you're right:
`<font-format>= [<string> | collection | embedded-opentype | opentype  | svg | truetype | woff | woff2 ]`

vitorroriz committed:
@font-face src format() should parse valid unsupported keywords
https://bugs.webkit.org/show_bug.cgi?id=259144
rdar://112135869

Reviewed by Tim Nguyen.

A valid format() keyword (ident) argument should be parsed even if the represented
format is not supported by the engine. The engine should reject loading
the font for unsupported formats but this should happen only at
loading time and not at parsing time.

The reason why we were rejecting it before, at parsing time, is for some
inconsistency in different parts of the spec. That
will be fixed soon by the CSSWG, see: https://github.com/w3c/csswg-drafts/issues/8793

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-format-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::isFontFormatKeywordValid):
(WebCore::CSSPropertyParserHelpers::consumeFontFormat):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::consumeSupportsFontFormatFunction):

Canonical link: https://commits.webkit.org/266043@main
